types/mempool/mempool.go

```diff
@@ -0,0 +1,254 @@
+package mempool
+
+import (
+	"fmt"
+	"math"
+
+	huandu "github.com/huandu/skiplist"
+
+	"github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/auth/signing"
+)
+
+// Tx we define an app-side mempool transaction interface that is as
+// minimal as possible, only requiring applications to define the size of the
+// transaction to be used when reaping and getting the transaction itself.
+// Interface type casting can be used in the actual app-side mempool implementation.
+type Tx interface {
+	types.Tx
+
+	// Size returns the size of the transaction in bytes.
+	Size() int64
+}
+
+type Mempool interface {
+	// Insert attempts to insert a Tx into the app-side mempool returning
+	// an error upon failure.
+	Insert(types.Context, Tx) error
+
+	// Select returns the next set of available transactions from the app-side
+	// mempool, up to maxBytes or until the mempool is empty. The application can
+	// decide to return transactions from its own mempool, from the incoming
+	// txs, or some combination of both.
+	Select(txs [][]byte, maxBytes int64) ([]Tx, error)
+
+	// CountTx returns the number of transactions currently in the mempool.
+	CountTx() int
+
+	// Remove attempts to remove a transaction from the mempool, returning an error
+	// upon failure.
+	Remove(Tx) error
+}
+
+type ErrTxNotFound struct {
+	error
+}
+
+type Factory func() Mempool
+
+var _ Mempool = (*defaultMempool)(nil)
+
+// defaultMempool is the SDK's default mempool implementation which stores txs in a partially ordered set
+// by 2 dimensions; priority, and sender-nonce.  Internally it uses one priority ordered skip list and one skip list
+// per sender ordered by nonce (sender-nonce).  When there are multiple txs from the same sender, they are not
+// always comparable by priority to other sender txs and must be partially ordered by both sender-nonce and priority.
+type defaultMempool struct {
```

Comment:
So it seems I had an initial misunderstanding in how this would be implemented. I thought we'd have two indexes (skip lists) -- one indexes by priority and the other by nonce. Why does the priority index need nonce, sender and all this other stuff?

Reply:
As written there is one index ordered by priority, and one index for each sender ordered by nonce.

The priority index carries sender and nonce in the key to resolve priority collisions, otherwise two tx with the same priority would overwrite each other.  So sender-nonce is part of the skip list key score calculation. In fact sender-nonce is used to uniquely identify a tx in the pool.